fix: upgrading celery version and fixing tests.

awais786 opened the pull request:
https://github.com/edx/edx-arch-experiments/issues/330

**What's new in Celery 5.3**
https://github.com/celery/celery/blob/3f965ebb9321c982f229429eb002ec23f114aa7d/docs/whatsnew-5.3.rst#deprecate-pytz-and-use-zoneinfo

**PR** https://github.com/celery/celery/pull/8159

**Upgrade requirements PR**

**List of packages in the PR without any issue.</br>** 
 - `celery` changes from `5.2.7` to `5.3.0`
- `django-cors-headers` changes from `4.0.0` to `4.1.0`
- `edx-sga` changes from `0.21.1` to `0.22.0`
- `kombu` changes from `5.2.4` to `5.3.0`
- `ora2` changes from `5.0.2` to `5.0.3`


**These Packages need manual review..</br>** 
 - **[REMOVED]**  `backports-zoneinfo` (`0.2.1`) removed from the requirements
- **[NEW]**  `backports-zoneinfo[tzdata]` (`0.2.1`) added to the requirements
- **[MAJOR]** `billiard` changes from `3.6.4.0` to `4.1.0`
- **[NEW]**  `tzdata` (`2023.3`) added to the requirements

Note: We need to check whether this change breaks actual functionality ?

https://zubairshakoorarbisoft.sandbox.edx.org/ Sandbox looks fine.

awais786 committed:
chore: upgrading celery version and fixing tests.

edx-requirements-bot committed:
chore: Updating Python Requirements (#32469)

zubairshakoorarbisoft committed:
fix: fixed unit test by changing date comparisons

zubairshakoorarbisoft committed:
fix: timestamps replaced with datetime object in tests

zubairshakoorarbisoft committed:
fix: notify creds tests fixes applied

zubairshakoorarbisoft committed:
fix: applied fixes to updated exceptions

awais786 committed:
fix: fixing traceback.

awais786 commented on lms/djangoapps/instructor_task/models.py:
```diff
@@ -178,11 +178,11 @@ def create_output_for_failure(exception, traceback_string):
                 else:
                     # truncate the traceback:
-                    task_progress['traceback'] = traceback_string[:-(too_long + len(tag))] + tag
+                    task_progress['traceback'] = traceback_string[(too_long + len(tag)):] + tag
```
Now celery is sending complete traceback and by trimming from end it removes the exception. So removes the initial part instead.

zubairshakoorarbisoft committed:
fix: message assertiion changed for failing tests

zubairshakoorarbisoft committed:
fix: short message assertion changed

zubairshakoorarbisoft committed:
fix: traceback key issue fixed in tests

zubairshakoorarbisoft committed:
fix: backttrace key changedd

rgraber reviewed (commented):
Does this mean we're losing all the error classes and will now just get "ExceptionWithTraceback" for everything? That seems like it will make error triage hard.

dianakhuang reviewed (approved):
Overall, I think this looks okay, but I think we would want to be extra careful to monitor for potential datetime issues on deploy to production

awais786 committed:
Merge branch 'master' into fixing-celery-tz-issue

github-actions[bot] committed:
feat: Upgrade Python dependency celery (#32672)

Commit generated by workflow `openedx/edx-platform/.github/workflows/upgrade-one-python-dependency.yml@refs/heads/master`

Co-authored-by: awais786 <awais786@users.noreply.github.com>

zubairshakoorarbisoft commented:
@rgraber thanks for having eyes on my PR, let me explain you the whole things in detail:

Previously with celery version `5.2.1` we were getting `NotImplementedError("Problem's definition does not support rescoring.")` in exception and `Traceback (most recent call last):\n  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/celery/app/trace.py", line 451, in trace_task\n    R = retval = fun(*args, **kwargs)\n  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edx_django_utils/monitoring/internal/code_owner/utils.py", line 193, in new_function\n    return wrapped_function(*args, **kwargs)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks.py", line 82, in rescore_problem\n    return run_main_task(entry_id, visit_fcn, action_name)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/runner.py", line 120, in run_main_task\n    task_progress = task_fcn(entry_id, course_id, task_input, action_name)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/module_state.py", line 96, in perform_module_state_update\n    update_status = update_fcn(block, module_to_update, task_input)\n  File "/usr/lib/python3.8/contextlib.py", line 75, in inner\n    return func(*args, **kwds)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/module_state.py", line 171, in rescore_problem_module_state\n    instance.rescore(only_if_higher=task_input[\'only_if_higher\'])\n  File "/edx/app/edxapp/edx-platform/xmodule/capa_block.py", line 2150, in rescore\n    raise NotImplementedError(_("Problem\'s definition does not support rescoring."))\nNotImplementedError: Problem\'s definition does not support rescoring.\n` in `traceback_string` and in `create_output_for_failure` function at `lms.djangoapps.instructor_task.models.py` we were simply getting error message from exception like `str(exception)`, but when we switched to celery `5.3.1`, we were started getting `ExceptionWithTraceback` in exception so we have the same traceback string in both of the variables (`exception, traceback_string`) passing into `create_output_for_failure` and based on the algorithm written in this function, if the length of `json_output` or you can say `traceback_string` they were striping the string and extracting error message from the start of string but now,`ExceptionWithTraceback` length has been changed so we are stripping exception string from bottom instead and removing the initial part.

**For Example:**

**1. The whole traceback string getting in both variables:**
```Traceback (most recent call last):\n  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/celery/app/trace.py", line 451, in trace_task\n    R = retval = fun(*args, **kwargs)\n  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edx_django_utils/monitoring/internal/code_owner/utils.py", line 193, in new_function\n    return wrapped_function(*args, **kwargs)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks.py", line 82, in rescore_problem\n    return run_main_task(entry_id, visit_fcn, action_name)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/runner.py", line 120, in run_main_task\n    task_progress = task_fcn(entry_id, course_id, task_input, action_name)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/module_state.py", line 96, in perform_module_state_update\n    update_status = update_fcn(block, module_to_update, task_input)\n  File "/usr/lib/python3.8/contextlib.py", line 75, in inner\n    return func(*args, **kwds)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/module_state.py", line 171, in rescore_problem_module_state\n    instance.rescore(only_if_higher=task_input[\'only_if_higher\'])\n  File "/edx/app/edxapp/edx-platform/xmodule/capa_block.py", line 2150, in rescore\n    raise NotImplementedError(_("Problem\'s definition does not support rescoring."))\nNotImplementedError: Problem\'s definition does not support rescoring.\n```

**2. The way it was stripping with old celery version, when we were not getting ExceptionWithTraceback, that's why error message getting missed, it was actually in the bottom of the traceback:**
```Traceback (most recent call last):\n  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/celery/app/trace.py", line 451, in trace_task\n    R = retval = fun(*args, **kwargs)\n  File "/edx/app/edxapp/venvs/edxapp/lib/python3.8/site-packages/edx_django_utils/monitoring/internal/code_owner/utils.py", line 193, in new_function\n    return wrapped_function(*args, **kwargs)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks.py", line 82, in rescore_problem\n    return run_main_task(entry_id, visit_fcn, action_name)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/runner.py", line 120, in run_main_task\n    task_progress = task_fcn(entry_id, course_id, task_input, action_name)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/module_state.py", line 96, in perform_module_st...```

**3. After making our changes in stripping, getting string from the end where the error exists for assertions:**
```ask/tasks_helper/runner.py", line 120, in run_main_task\n    task_progress = task_fcn(entry_id, course_id, task_input, action_name)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/module_state.py", line 96, in perform_module_state_update\n    update_status = update_fcn(block, module_to_update, task_input)\n  File "/usr/lib/python3.8/contextlib.py", line 75, in inner\n    return func(*args, **kwds)\n  File "/edx/app/edxapp/edx-platform/lms/djangoapps/instructor_task/tasks_helper/module_state.py", line 171, in rescore_problem_module_state\n    instance.rescore(only_if_higher=task_input[\'only_if_higher\'])\n  File "/edx/app/edxapp/edx-platform/xmodule/capa_block.py", line 2150, in rescore\n    raise NotImplementedError(_("Problem\'s definition does not support rescoring."))\nNotImplementedError: Problem\'s definition does not support rescoring.\n"""...```

### Conclusion
so, the conclusion is that, the function/algorithm first try to get the error message for assertion from `task_progress = {'exception': type(exception).__name__, 'message': str(exception)}` means directly from the exception like `'message': str(exception)` and at the end based on the length of traceback they are shortening the traceback to prepare `task_progress['message']` explicitly, and in new celery version, we are always getting `ExceptionWithTraceback` in exception that's why message is not setting directly and string manipulation happening in the algorithm and I made the fix to fetch the message correctly from the traceback string.

Please let me know, if you still and have any questions.

awais786 committed:
Merge branch 'master' into fixing-celery-tz-issue

zubairshakoorarbisoft commented:
I've created sandbox for this PR and tested it, everything looks good
